### PR TITLE
Refactoring of request logging and feature  #508

### LINKED ISSUE #508: I want the host log to show if the requested host was found
Many user that use filtering DNS server have problems debugging why a certain app is not working because Blokada only shows if it blocked the request. Right now it's not possible to see if the request succeeded or maybe got blocked by the DNS server.

It would be nice if Blokada would check the answer for all allowed DNS querys and would mark every host that was not found.
The entry in the host log should look different from blocked hosts to make it easier to differentiate.

### PR DESCRIPTION
**Pull request by ps100000**
Fixes #508

There where a lot of different pieces of code which all handled saving requests in a slightly different way. I replaced all of them with one central class which handles everything connected to logging.
I also tried to document the inner working of this new logging code.

### CODE REVIEW

File: app/src/full/kotlin/tunnel/DnsProxy.kt [hunk 1]
```diff
@@ -59,25 +61,37 @@ internal class DnsProxy(
             w("failed reading DNS message", e)
             return
         }
+        dnsMessage.question.name
```
Comment (kar) [inline app/src/full/kotlin/tunnel/DnsProxy.kt:64]: What does it mean?

Comment (ps100000) [reply app/src/full/kotlin/tunnel/DnsProxy.kt:64]: Looks like I overlooked some code fragment from debugging. I'll wait for the full review and remove it after that. :)

File: app/src/full/kotlin/tunnel/DnsProxy.kt [hunk 2]
```diff
@@ -59,25 +61,37 @@ internal class DnsProxy(
...
-            emit(TunnelEvents.REQUEST, Request(host, blocked = true))
+            RequestLog.add(ExtendedRequest(host, blocked = true))
         }
     }
 
     override fun toDevice(response: ByteArray, length: Int, originEnvelope: Packet?) {
         originEnvelope as IpPacket
+
+        if (isAnswerValid(response)){
+            v("easytosearchstring1 valid Answer")
```
Comment (kar) [inline app/src/full/kotlin/tunnel/DnsProxy.kt:86]: Testing log message?

File: app/src/full/kotlin/tunnel/DnsProxy.kt [hunk 3]
```diff
@@ -59,25 +61,37 @@ internal class DnsProxy(
...
+        }else{
+            val responseName = Name(response.sliceArray(12 until response.size)).canonicalize().toString(true)
```
Comment (kar) [inline app/src/full/kotlin/tunnel/DnsProxy.kt:88]: Why 12?

File: app/src/full/kotlin/tunnel/DnsProxy.kt [hunk 4]
```diff
@@ -59,25 +61,37 @@ internal class DnsProxy(
...
+
+            v("easytosearchstring1 INVALID ANSWER!\nUpdated entry:",
```
Comment (kar) [inline app/src/full/kotlin/tunnel/DnsProxy.kt:90]: string

File: app/src/main/res/values-ar-rSA/strings_panel.xml [hunk 5]
```diff
@@ -25,7 +25,7 @@
     <string name="panel_domain_forwarded_toast">تَمَّت الإضافة إلى القائمة البيضاء.</string>
     <!-- When a domain is blocked -->
     <string name="panel_domain_blocked">&lt;b&gt;%s&lt;/b&gt; محظور</string>
-    <string name="panel_domain_blocked_desc">تَمّ حَظْر هذا الموقع. للسماح به في المستقبل ، استخدم الزرّ أدناه.</string>
+    <string name="panel_domain_blocked_normal_desc">تَمّ حَظْر هذا الموقع. للسماح به في المستقبل ، استخدم الزرّ أدناه.</string>
```
Comment (kar) [inline app/src/main/res/values-ar-rSA/strings_panel.xml:28]: Please consider not changing existing string keys in the future, as it makes all translators have to manually update this translations everywhere in the world.

File: app/src/tun-local/kotlin/tunnel/Model.kt [hunk 6]
```diff
@@ -110,17 +111,60 @@ data class BlockaConfig(
...
+    ALLOWED_APP_UNKNOWN, // allowed app unknown
+    ALLOWED_APP_KNOWN    // allowed app known ( future use in firewall )
+}
+
+
+data class ExtendedRequest(
+        override val domain: String,
+        override val time: Date = Date(),
+        var state: RequestState = RequestState.ALLOWED_APP_UNKNOWN,
+        var ip: InetAddress? = null, // for future use in firewall
```
Comment (kar) [inline app/src/tun-local/kotlin/tunnel/Model.kt:148]: interesting!

File: app/src/tun-local/kotlin/tunnel/Persistence.kt [hunk 7]
```diff
@@ -93,24 +93,24 @@ class FiltersPersistence {
 }
 
 class RequestPersistence(
```
Comment (kar) [inline app/src/tun-local/kotlin/tunnel/Persistence.kt:95]: Could you describe in a few sentences how the batch loading works in the new code? What's the algorithm?

File: app/src/ui-blokada/kotlin/core/bits/menu/adblocking/HostsLogVB.kt [hunk 8]
```diff
@@ -24,59 +24,58 @@ class HostsLogVB(
...
-                items.add(0, dash)
-                view?.add(dash, 3)
-                firstItem = it
-                onSelectedListener(null)
-            }
-        }
-        Unit
-    }
+    private val requestUpdate = { update: RequestUpdate ->
+                                    if (searchString.isEmpty() || update.newState.domain.contains(searchString.toLowerCase())) {
```
Comment (kar) [inline app/src/ui-blokada/kotlin/core/bits/menu/adblocking/HostsLogVB.kt:32]: some strange padding?

File: docs/RequestLogging.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
+# The RequestLog
```
Comment (kar) [inline docs/RequestLogging.md:1]: Great description!

**Comment by kar** [comment]
It seems the playstore build fails: (./gradlew iGB)

```
e: /Users/kar/code/github.com/blokadaorg/blokada/app/src/google/kotlin/flavor/Module.kt: (14, 17): Unresolved reference: updateControllswitchWidgets
e: /Users/kar/code/github.com/blokadaorg/blokada/app/src/google/kotlin/flavor/Module.kt: (16, 13): Unresolved reference: updateControllswitchWidgets
e: /Users/kar/code/github.com/blokadaorg/blokada/app/src/google/kotlin/flavor/Module.kt: (18, 15): Unresolved reference: tunnelDropCount
```

File: app/src/full/kotlin/tunnel/DnsProxy.kt (hunk 2 above)
Comment (ps100000) [reply app/src/full/kotlin/tunnel/DnsProxy.kt:86]: Yes. I'll remove it.

File: app/src/main/res/values-ar-rSA/strings_panel.xml (hunk 5 above)
Comment (ps100000) [reply app/src/main/res/values-ar-rSA/strings_panel.xml:28]: should I change it back?

File: app/src/full/kotlin/tunnel/DnsProxy.kt (hunk 3 above)
Comment (ps100000) [reply app/src/full/kotlin/tunnel/DnsProxy.kt:88]: Added a comment:
// the first 12 bytes of response contain the DNS-header and are cut of. The header
// should be followed by a resource record which starts with the name that was requested.

File: app/src/full/kotlin/tunnel/DnsProxy.kt (hunk 4 above)
Comment (ps100000) [reply app/src/full/kotlin/tunnel/DnsProxy.kt:90]: removed

**Pull request merged**